### PR TITLE
Mapping validation

### PR DESCRIPTION
**Pull request by ulugbek**

### CODE REVIEW

File: xtraplatform-feature-provider-sql/src/main/java/de/ii/xtraplatform/feature/provider/sql/infra/db/SqlTypeInfoValidator.java [hunk 1]
```diff
@@ -89,12 +95,45 @@ public SqlTypeInfoValidator(ConnectionInfoSql connectionInfo, ClassLoader classL
...
+                if (!tableExists(relation.getTargetContainer())) {
+                    errors.add(String.format("table '%s' does not exist", relation.getTargetContainer()));
+                }
 
                 //TODO: columnExists for relation.getSourceField(), relation.getSourceSortKey() and relation.getTargetField()
+                if (relation.getJunction().isPresent() && !tableExists(relation.getJunction().get())) {
+                    errors.add(String.format("table '%s' does not exist", relation.getJunction().get()));
+                }
+
+                if (!columnExists(relation.getSourceField(), relation.getSourceSortKey())) {
```
Comment (azahnen) [inline xtraplatform-feature-provider-sql/src/main/java/de/ii/xtraplatform/feature/provider/sql/infra/db/SqlTypeInfoValidator.java:111]: The second parameter should be `relation.getSourceContainer()`.

File: xtraplatform-feature-provider-sql/src/main/java/de/ii/xtraplatform/feature/provider/sql/infra/db/SqlTypeInfoValidator.java [hunk 2]
```diff
@@ -89,12 +95,45 @@ public SqlTypeInfoValidator(ConnectionInfoSql connectionInfo, ClassLoader classL
...
+                    errors.add(String.format("column '%s' in table '%s' does not exist", relation.getSourceField(), relation.getSourceSortKey()));
+                }
+
+                if (!columnExists(relation.getTargetField(), relation.getSourceSortKey())) {
```
Comment (azahnen) [inline xtraplatform-feature-provider-sql/src/main/java/de/ii/xtraplatform/feature/provider/sql/infra/db/SqlTypeInfoValidator.java:115]: The second parameter should be `relation.getTargetContainer()`.

File: xtraplatform-feature-provider-sql/src/main/java/de/ii/xtraplatform/feature/provider/sql/infra/db/SqlTypeInfoValidator.java [hunk 3]
```diff
@@ -89,12 +95,45 @@ public SqlTypeInfoValidator(ConnectionInfoSql connectionInfo, ClassLoader classL
...
 
                 //TODO: columnExists for relation.getSourceField(), relation.getSourceSortKey() and relation.getTargetField()
+                if (relation.getJunction().isPresent() && !tableExists(relation.getJunction().get())) {
+                    errors.add(String.format("table '%s' does not exist", relation.getJunction().get()));
+                }
+
+                if (!columnExists(relation.getSourceField(), relation.getSourceSortKey())) {
+                    errors.add(String.format("column '%s' in table '%s' does not exist", relation.getSourceField(), relation.getSourceSortKey()));
+                }
+
```
Comment (azahnen) [inline xtraplatform-feature-provider-sql/src/main/java/de/ii/xtraplatform/feature/provider/sql/infra/db/SqlTypeInfoValidator.java:114]: Repeat check with column `relation.getSourceSortKey()` and table `relation.getSourceContainer()`.

File: xtraplatform-feature-provider-sql/src/main/java/de/ii/xtraplatform/feature/provider/sql/infra/db/SqlTypeInfoValidator.java [hunk 4]
```diff
@@ -89,12 +95,45 @@ public SqlTypeInfoValidator(ConnectionInfoSql connectionInfo, ClassLoader classL
...
+                if (!columnExists(relation.getTargetField(), relation.getSourceSortKey())) {
+                    errors.add(String.format("column '%s' in table '%s' does not exist", relation.getTargetField(), relation.getSourceSortKey()));
+                }
 
                 //TODO: columnExists for relation.getJunctionSource() and relation.getJunctionTarget() if present
+                if (relation.getJunctionSource().isPresent() && relation.getJunctionTarget().isPresent()) {
+                    if (!columnExists(relation.getJunctionSource().get(), relation.getJunctionTarget().get())) {
```
Comment (azahnen) [inline xtraplatform-feature-provider-sql/src/main/java/de/ii/xtraplatform/feature/provider/sql/infra/db/SqlTypeInfoValidator.java:121]: The second parameter should be `relation.getJunction().get()`.

File: xtraplatform-feature-provider-sql/src/main/java/de/ii/xtraplatform/feature/provider/sql/infra/db/SqlTypeInfoValidator.java [hunk 5]
```diff
@@ -89,12 +95,45 @@ public SqlTypeInfoValidator(ConnectionInfoSql connectionInfo, ClassLoader classL
...
+                if (!columnExists(relation.getSourceField(), relation.getSourceSortKey())) {
+                    errors.add(String.format("column '%s' in table '%s' does not exist", relation.getSourceField(), relation.getSourceSortKey()));
+                }
+
+                if (!columnExists(relation.getTargetField(), relation.getSourceSortKey())) {
+                    errors.add(String.format("column '%s' in table '%s' does not exist", relation.getTargetField(), relation.getSourceSortKey()));
+                }
 
                 //TODO: columnExists for relation.getJunctionSource() and relation.getJunctionTarget() if present
+                if (relation.getJunctionSource().isPresent() && relation.getJunctionTarget().isPresent()) {
```
Comment (azahnen) [inline xtraplatform-feature-provider-sql/src/main/java/de/ii/xtraplatform/feature/provider/sql/infra/db/SqlTypeInfoValidator.java:120]: The second presence check should be `relation.getJunction().isPresent()`.

File: xtraplatform-feature-provider-sql/src/main/java/de/ii/xtraplatform/feature/provider/sql/infra/db/SqlTypeInfoValidator.java [hunk 6]
```diff
@@ -89,12 +95,45 @@ public SqlTypeInfoValidator(ConnectionInfoSql connectionInfo, ClassLoader classL
...
+                    if (!columnExists(relation.getJunctionSource().get(), relation.getJunctionTarget().get())) {
+                        errors.add(String.format("column '%s' in table '%s' does not exist",
+                                relation.getJunctionSource().get(), relation.getJunctionTarget().get()));
+                    }
+                }
```
Comment (azahnen) [inline xtraplatform-feature-provider-sql/src/main/java/de/ii/xtraplatform/feature/provider/sql/infra/db/SqlTypeInfoValidator.java:126]: Repeat check with column `relation.getJunctionTarget()` and table `relation.getJunction()`.

File: xtraplatform-feature-provider-sql/src/main/java/de/ii/xtraplatform/feature/provider/sql/infra/db/SqlTypeInfoValidator.java [hunk 7]
```diff
@@ -89,12 +95,45 @@ public SqlTypeInfoValidator(ConnectionInfoSql connectionInfo, ClassLoader classL
...
 
                 //TODO: isColumnUnique for relation.getSourceSortKey()
+                if (!isColumnUnique(relation.getSourceField(), relation.getSourceSortKey())) {
```
Comment (azahnen) [inline xtraplatform-feature-provider-sql/src/main/java/de/ii/xtraplatform/feature/provider/sql/infra/db/SqlTypeInfoValidator.java:128]: The first parameter should be `relation.getSourceSortKey()`, the second `relation.getSourceContainer()`.

File: xtraplatform-feature-provider-sql/src/main/java/de/ii/xtraplatform/feature/provider/sql/infra/db/SqlTypeInfoValidator.java [hunk 8]
```diff
@@ -89,12 +95,45 @@ public SqlTypeInfoValidator(ConnectionInfoSql connectionInfo, ClassLoader classL
...
+                    errors.add(String.format("column %s in table %s is not a primary key or a unique index for the column does not exist",
+                            relation.getSourceField(), relation.getSourceSortKey()));
+                }
+
+                if (!isColumnUnique(relation.getTargetField(), relation.getSourceSortKey())) {
```
Comment (azahnen) [inline xtraplatform-feature-provider-sql/src/main/java/de/ii/xtraplatform/feature/provider/sql/infra/db/SqlTypeInfoValidator.java:133]: Remove this check.

File: xtraplatform-feature-provider-sql/src/main/java/de/ii/xtraplatform/feature/provider/sql/infra/db/SqlTypeInfoValidator.java [hunk 9]
```diff
@@ -61,25 +63,29 @@ public SqlTypeInfoValidator(ConnectionInfoSql connectionInfo, ClassLoader classL
...
+                    errors.add(String.format("column %s in table %s is not a primary key or a unique index for the column does not exist",
+                            attributesContainer.getSortKey(), attributesContainer.getName()));
                 }
 
-                if (attribute.isSpatial()) {
-                    //TODO: isColumnSpatial( attribute.getName(), attributesContainer.getName())
-                    // -> check that column is of type geometry
+                if (attribute.isSpatial() && !isColumnSpatial(attribute.getName(), attributesContainer.getName())) {
+                    errors.add(String.format("column %s in table %s is not of type geometry", attribute.getName(), attributesContainer.getName()));
                 }
```
Comment (azahnen) [inline xtraplatform-feature-provider-sql/src/main/java/de/ii/xtraplatform/feature/provider/sql/infra/db/SqlTypeInfoValidator.java:89]: Add similar check with `attribute.isTemporal()` (maybe you have to pull, I added this later) and `isColumnTemporal` (to implement, see interactive-instruments/ldproxy#260).

File: xtraplatform-feature-provider-sql/src/main/java/de/ii/xtraplatform/feature/provider/sql/infra/db/SqlTypeInfoValidator.java [hunk 10]
```diff
@@ -103,13 +142,31 @@ public SqlTypeInfoValidator(ConnectionInfoSql connectionInfo, ClassLoader classL
...
+                .anyMatch(c -> c.getName().equals(name));
+    }
+
+    private boolean isColumnUnique(String name, String table) {
+        Optional<Column> column = tables.stream()
+                .filter(t -> t.getName().equals(table))
+                .flatMap(t -> t.getColumns().stream())
+                .filter(c -> c.getName().equals(name))
+                .findFirst();
+        return column.isPresent() && (column.get().isPartOfPrimaryKey() || column.get().isPartOfUniqueIndex());
```
Comment (azahnen) [inline xtraplatform-feature-provider-sql/src/main/java/de/ii/xtraplatform/feature/provider/sql/infra/db/SqlTypeInfoValidator.java:161]: I think `isPartOf` might not be strict enough, the index should contain only this column.

For primary key you can check that on the table: 

```
table.getPrimaryKey().getColumns().size() == 1 && table.getPrimaryKey().getColumns().get(0).getReferencedColumn().equals(column)
```

For unique indexes you would have to iterate over `table.getIndexes()`, filter by `isUnique()` and then check like above that `getColumns()` only contains the given `column`.

**Review by azahnen — CHANGES_REQUESTED** [review]

**Comment by ulugbek** [comment]
All the issues have been fixed.

**Review by azahnen — APPROVED** [review]

**Review by azahnen — APPROVED** [review]

**Pull request merged**